reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,417 @@
+                for (LocalEstimate unaryEstimate : unaryEstimates) {
+                    if (!unaryCostCover.containsKey(unaryEstimate.variables.get(0))) {
+                        unaryCostCover.put(unaryEstimate.variables.get(0), new CostCover(unaryEstimate.answerEstimate(this, new HashSet<>(unaryEstimate.variables))));
+                    }
+                }
+            }
+
+            assert multivarEstimates == null;
+            multivarEstimates = new HashMap<>();
+            resolvables().forEachRemaining(resolvable -> multivarEstimates.put(resolvable, new ArrayList<>()));
+
+            //TODO: How does traversal handle type Variables?
+
+            resolvables().filter(Resolvable::isRetrievable).map(Resolvable::asRetrievable).forEachRemaining(retrievable -> {
+                Set<Concludable> concludablesInRetrievable = ResolvableConjunction.of(retrievable.pattern()).positiveConcludables();
+                iterate(concludablesInRetrievable).forEachRemaining(concludable -> {
+                    multivarEstimates.get(retrievable).addAll(computeEstimatesFromConcludable(concludable));
+                });
+            });
+
+            iterate(resolvables()).filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                multivarEstimates.get(concludable).addAll(computeEstimatesFromConcludable(concludable));
+            });
+
+            if (this.negatedsCost == -1) {
+                this.negatedsCost = resolvables().filter(Resolvable::isNegated).map(Resolvable::asNegated)
+                        .flatMap(negated -> iterate(negated.disjunction().conjunctions()))
+                        .map(negatedConjunction -> {
+                            costEstimator.registerConjunction(negatedConjunction);
+                            return costEstimator.estimateAllAnswers(negatedConjunction);
+                        }).reduce(0L, Long::sum);
+            }
+
+            // Can we prune the multiVarEstimates stored?
+            // Not if we want to order-resolvables. Else, yes based on queryableVariables.
+
+            Map<Variable, CostCover> fullCostCover = computeCostCover(iterate(multivarEstimates.values()).flatMap(Iterators::iterate).toList(), allVariables());
+            this.fullAnswerCount = CostCover.costToCover(allVariables(), fullCostCover) + this.negatedsCost;
+        }
+
+        private List<LocalEstimate> computeEstimatesFromConcludable(Concludable concludable) {
+            if (concludable.isHas()) {
+                return list(estimatesFromHasEdges(concludable.asHas()));
+            } else if (concludable.isRelation()) {
+                return list(estimatesFromRolePlayersAssumingEvenDistribution(concludable.asRelation()));
+            } else {
+                return list();
+            }
+        }
+
+        private long computeInferredUnaryEstimates(ThingVariable generatedVariable) {
```

Comment:
For attribute types:, adds one to the unary estimate for each ExplicitHas rule concluding the type (Even if it already occurs in the data)
For relation-types: Sums across each rule concluding a matching type, the estimated answers to the rule (i.e., to the condition of the rule projected to the variables in the head of the rule)